### PR TITLE
iOS platformLayers call onOutsidePointerEvent in hitTest

### PR DESCRIPTION
**Pull request by dima-avdeev-jb**
backgroundView should pass touches to other views for `focused=false` and intercept touches for `focused=true`. We can do it only in overriden `hitTest` function. Only this function can correctly determine should backgroundView intercept or pass touches.
But we also can use `override fun touchesEnded` in case of `focused==true` and call `onOutsidePointerEvent?.invoke(PointerEventType.Release)` there

Also I made a workaround to not call `onOutsidePointerEvent` twice. We can save `previousSuccessHitTestTimestamp` and compare it in hitTest function. Tested it on simulator and real device - works as expected.

### CODE REVIEW

**Comment by dima-avdeev-jb** [comment]
@igordmn Can you please check this branch on your iOS simulator?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt [hunk 1]
```diff
@@ -62,22 +62,26 @@ internal class UIViewComposeSceneLayer(
...
-
-        override fun pointInside(point: CValue<CGPoint>, withEvent: UIEvent?): Boolean {
-            //TODO pass invoke(true) on touch up event only when this touch event begins outside of layer bounds.
-            // Also it should be only one touch event (not multitouch with 2 and more touches).
-            // In other cases pass invoke(false)
-            onOutsidePointerEvent?.invoke(PointerEventType.Press)
-            onOutsidePointerEvent?.invoke(PointerEventType.Release)
-            return focusable
-        }
+        private var previousSuccessHitTestTimestamp: Double? = null
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt:71]: You don't need to add any timestamps here

**Review by MatkovIvan — CHANGES_REQUESTED** [review]
Disagree this this
Passing touches can be done manully it it's not possible to filter by platform API.

I mean you DO have all references for all our layers and you CAN manually call methods like `touchesBegan` for other views

**Comment by elijah-semyonov** [comment]
> Disagree this this Passing touches can be done manully it it's not possible to filter by platform API.
> 
> I mean you DO have all references for all our layers and you CAN manually call methods like `touchesBegan` for other views

It's a bad practice to call `touchesBegan` on other views manually. You must uphold correct invariant of correct order of `touchesBegan`->`touchesMoved`->`touchesEnded` or `touchesCancelled` in this case which is much trickier, than it seems.

Much better approach is to define custom UIGestureRecognizer's for this case.
Let's discuss it later on the meeting.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt [hunk 2]
```diff
@@ -62,22 +62,26 @@ internal class UIViewComposeSceneLayer(
...
+        override fun hitTest(point: CValue<CGPoint>, withEvent: UIEvent?): UIView? =
+            if (
+                mediator.hitTestInteractionView(point, withEvent) == null &&
+                super.hitTest(point, withEvent) == this
+            ) {
+                if (previousSuccessHitTestTimestamp != withEvent?.timestamp) {
+                    // It is workaround to call onOutsidePointerEvent just once for one touch event.
+                    previousSuccessHitTestTimestamp = withEvent?.timestamp
+                    onOutsidePointerEvent?.invoke(PointerEventType.Press)
+                    onOutsidePointerEvent?.invoke(PointerEventType.Release)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]: Please move this call into `touchEnded` like we discussed. It's not required to call if it's not focusable

Comment (dima-avdeev-jb) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]: `PointerEventType.Release` moved to `touchesEnded`

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt (hunk 1 above)
Comment (dima-avdeev-jb) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt:71]: Discussed it - we still need `previousSuccessHitTestTimestamp` to call `onOutsidePointerEvent?.invoke(PointerEventType.Press)` just once on one touch

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt [hunk 3]
```diff
@@ -185,6 +185,9 @@ internal class ComposeSceneMediator(
         coroutineContext: CoroutineContext
     ) -> ComposeScene,
 ) {
+    fun hitTestInteractionView(point: CValue<CGPoint>, withEvent: UIEvent?): UIView? =
+        interactionView.hitTest(point, withEvent)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.uikit.kt]: Let's keep methods after properties. Just move it below

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt [hunk 4]
```diff
@@ -62,22 +67,47 @@ internal class UIViewComposeSceneLayer(
...
         frame = CGRectZero.readValue()
     ) {
-        init {
-            translatesAutoresizingMaskIntoConstraints = false
-            rootView.addSubview(this)
-            NSLayoutConstraint.activateConstraints(
-                getConstraintsToFillParent(this, rootView)
-            )
+
+        var previousSuccessHitTestTimestamp: Double? = null
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]: private

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt [hunk 5]
```diff
@@ -62,22 +67,47 @@ internal class UIViewComposeSceneLayer(
...
+        /**
+         * touchesEnded calls only when focused == true
+         */
+        override fun touchesEnded(touches: Set<*>, withEvent: UIEvent?) {
+            val touch = touches.firstOrNull() as? UITouch
+            val locationInView = touch?.locationInView(this)
+            if (locationInView != null) {
+                val offset = locationInView.useContents { toDpOffset() }
+                val contains = boundsInWindow.contains(offset.toOffset(density).round())
+                println("$boundsInWindow contains ${offset.toOffset(density).round()} is $contains")
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]:
```suggestion
```

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt [hunk 6]
```diff
@@ -62,22 +67,47 @@ internal class UIViewComposeSceneLayer(
...
+                super.hitTest(point, withEvent) == this
+            ) {
+                touchStartedOutside(withEvent)
+                if (focusable) {
+                    return this // block touches
+                }
+            }
+            return null // transparent for touches
         }
+
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]:
```suggestion
```

**Review by MatkovIvan — APPROVED** [review]

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt (hunk 5 above)
Comment (dima-avdeev-jb) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]: done

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt (hunk 6 above)
Comment (dima-avdeev-jb) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIViewComposeSceneLayer.uikit.kt]: done

**Pull request merged**